Tool & ToolManager, also DrawableGLObject with a DrawableAxes example

timoffex opened the pull request.

timoffex committed:
Scene and RenderableObject data structures

A RenderableObject contains just vertex data for now. A Scene contains
RenderableObjects. A shared_ptr<Scene> will be passed to the rendering
code (not yet committed).

timoffex committed:
added gitignore

Ignores everything in the QCreator-generated build…/ directory

timoffex committed:
Updated gitignore

Using https://github.com/github/gitignore/blob/master/Qt.gitignore
because whoever wrote this seems to know what they’re doing

timoffex committed:
Very basic rendering

Added button to MainWindow that opens the MeshView with a scene that
just contains a triangle. See MainWindow::openMeshView().

Added code to main.cpp to set the QSurfaceFormat default format. This
seems to be necessary for rendering OpenGL (at least on some systems
(?)).

Camera can be dragged in MeshView (in XY plane) using the mouse.

The rendered mesh is always purple.

Not yet tested with anything more complicated than a triangle.

timoffex committed:
Added to gitignore

Seems like the other one didn’t ignore this particular .user file for
some reason.

timoffex committed:
Merge branch 'master' into timoffex

timoffex committed:
Changed some std classes for Qt classes

std::vector -> QVector
std::shared_ptr -> QSharedPointer

timoffex committed:
Added Camera class

MeshView now uses perspective projection.

Click and drag with the left mouse button to rotate view. There is no
way to pan just yet.

timoffex committed:
Camera panning

Added camera panning. Hold shift when you start dragging to pan the
camera.

Commit:
base on tima's branch, added obj loading function, a bug fixed.

Commit:
base on tima's branch, added obj loading function, a bug fixed.

Commit:
remove .pro.user file

Commit:
remove .pro.user file

Commit:
fresh to death with onion breath

timoffex committed:
Merge pull request #3 from Bdtrotte/wswlymf

Wswlymf

Really nice work!

timoffex committed:
Merge branch 'master' into timoffex

timoffex committed:
Fix panning

Fixed a miscalculation in the panning code that made it misbehave a
little.

timoffex committed:
Fix panning

Fixed a miscalculation in the panning code that made it misbehave a
little.

timoffex committed:
DrawableGLObject and axes

Added the DrawableGLObject. This is any object that needs special
rendering. They can be added to a Scene just like RenderableObjects.

Note for future: I have noticed occasional crashes happening right when
the Show View button is pressed in the UI. These happen infrequently
and without a pattern.

timoffex committed:
DrawableGLObject and axes

Added the DrawableGLObject. This is any object that needs special
rendering. They can be added to a Scene just like RenderableObjects.

Note for future: I have noticed occasional crashes happening right when
the Show View button is pressed in the UI. These happen infrequently
and without a pattern.

timoffex committed:
Removed constructor from DrawableGLObject

It is an abstract class that should not be default-constructible.

timoffex committed:
Removed constructor from DrawableGLObject

It is an abstract class that should not be default-constructible.

timoffex committed:
Tool and ToolManager

Tool is an abstract class representing a UI tool (like a brush).

ToolManager represents a group of mutually exclusive tools, e.g.
“rectangle”, “line”, “erase”.

timoffex committed:
Tool and ToolManager

Tool is an abstract class representing a UI tool (like a brush).

ToolManager represents a group of mutually exclusive tools, e.g.
“rectangle”, “line”, “erase”.

timoffex committed:
Merged origin/master into timoffex

timoffex committed:
Added files to WallsAndHoles.pro (messed up in merge)

Bdtrotte commented on WallsAndHoles/toolmanager.cpp:
```diff
@@ -0,0 +1,37 @@
+}
+
+
+void ToolManager::registerTool(ToolP tool, QString name) {
```
Should return a QAction so it's easy to add to a tool bar

Bdtrotte commented on WallsAndHoles/tool.h:
```diff
@@ -0,0 +1,44 @@
+ * to function. A Brush tool should have some way of accessing the TileMap
+ * and some way of determining the visual parameters of the grid.
+ */
+class Tool {
```
Name abstract classes Abstract... so it is clear when implementing them

Bdtrotte commented on WallsAndHoles/toolmanager.cpp:
```diff
@@ -0,0 +1,37 @@
+#include "toolmanager.h"
```
In general the tool system will likely evolve in a lot of ways as we figure the rest of the system out, mainly in terms of using QActions and such

Bdtrotte reviewed (changes requested):
Didn't look through the 3D code too much, trust it's good!

As for the tool system, make the little changes I commented on for this pr, and we'll merge (you can go ahead and do it when ready) and as I said, build on the system as we go!

timoffex committed:
Camera zooming

MeshViewCamera class now has a wheelEvent() method too.

MeshViewCameraLikeBlender implements zooming using the wheelEvent().
Zoom by scrolling.

timoffex committed:
Added Abstract prefix to abstract classes

Tool -> AbstractTool
MeshViewCamera -> AbstractMeshViewCamera
DrawableGLObject -> AbstractDrawableGLObject

timoffex committed:
Amend to previous commit

Forgot to include AbstractDrawableGLObject

timoffex committed:
Better panning

Camera panning now considers zoom level.

timoffex committed:
Example Tool Usage, MeshView Container Window

MeshViewContainer class represents the window that contains the
MeshView. This is so that UI elements (such as a tool selection) can be
added.

The Blender-like camera now inherits from AbstractTool. See
MeshView::MeshView() for example code using the camera along with a
tool manager. Also see MeshViewContainer.

timoffex committed:
ToolManager returns QAction

ToolManager::registerTool() returns a QAction. I have not tested this
yet, but it should work.

Bdtrotte committed:
Merge branch 'master' into timoffex
